README.md
# lita-slack

[![Gem Version](https://badge.fury.io/rb/lita-slack.png)](http://badge.fury.io/rb/lita-slack)
[![Build Status](https://travis-ci.com/Shopify/lita-slack.svg?branch=master)](https://travis-ci.com/Shopify/lita-slack)

**lita-slack** is an adapter for [Lita](https://www.lita.io/) that allows you to use the robot with [Slack](https://slack.com/). The current adapter is not compatible with pre-1.0.0 versions, as it now uses Slack's [Real Time Messaging API](https://api.slack.com/rtm).

## Installation

Add **lita-slack** to your Lita instance's Gemfile:

```ruby
gem "lita-slack"
```

## Local Testing

```
bundle install
bundle exec rake test:unit
```

## Configuration

### Required attributes

- `token` (String) – The bot's Slack API token. Create a bot and get its token at https://my.slack.com/services/new/lita.

### Optional attributes

- `link_names` (Boolean) – Set to `true` to turn all Slack usernames in messages sent by Lita into links.
- `parse` (String) – Specify the parsing mode. See https://api.slack.com/docs/formatting#parsing_modes.
- `proxy` (String) – Specify a HTTP proxy URL. (e.g. "http://squid.example.com:3128")
- `unfurl_links` (Boolean) – Set to `true` to automatically add previews for all links in messages sent by Lita.
- `unfurl_media` (Boolean) – Set to `false` to prevent automatic previews for media files in messages sent by Lita.

**Note**: When using lita-slack, the adapter will overwrite the bot's name and mention name with the values set on the server, so `config.robot.name` and `config.robot.mention_name` will have no effect.

### config.robot.admins

Each Slack user has a unique ID that never changes even if their real name or username changes. To populate the `config.robot.admins` attribute, you'll need to use these IDs for each user you want to mark as an administrator. If you're using Lita version 4.1 or greater, you can get a user's ID by sending Lita the command `users find NICKNAME_OF_USER`.

### Example

```ruby
Lita.configure do |config|
  config.robot.adapter = :slack
  config.robot.admins = ["U012A3BCD"]

  config.adapters.slack.token = "abcd-1234567890-hWYd21AmMH2UHAkx29vb5c1Y"

  config.adapters.slack.link_names = true
  config.adapters.slack.parse = "full"
  config.adapters.slack.unfurl_links = false
  config.adapters.slack.unfurl_media = false
end
```

## Joining rooms

Lita will join your default channel after initial setup. To have it join additional channels or private groups, simply invite it to them via your Slack client as you would any normal user.

## Events

- `:connected` - When the robot has connected to Slack. No payload.
- `:disconnected` - When the robot has disconnected from Slack. No payload.
- `:slack_channel_created` - When the robot creates/updates a channel's or group's info, as directed by Slack. The payload has a single object, a `Lita::Slack::Adapters::SlackChannel` object, under the `:slack_channel` key.
- `:slack_reaction_added` - When a reaction has been added to a previous message. The payload includes `:user` (a `Lita::User` for the sender of the message in question), `:name` (the string name of the reaction added), `:item_user` (a `Lita::User` for the user that created the original item that has been reacted to), `:item` (a hash of raw data from Slack about the message), and `:event_ts` (a string timestamp used to identify the message).
- `:slack_reaction_removed` - When a reaction has been removed from a previous message. The payload is the same as the `:slack_reaction_added` message.
- `:slack_user_created` - When the robot creates/updates a user's info - name, mention name, etc., as directed by Slack. The payload has a single object, a `Lita::Slack::Adapters::SlackUser` object, under the `:slack_user` key.
- `:message_deleted` - When a message is deleted. Generates a payload that contains the deleted message's channel (under `"channel"`) and timestamp (`"deleted_ts"`)
- `:message_changed` - When a message is edited. Generates a payload that contains a message hash for the edited message under the `"message"` key.

## Chat service API

lita-slack supports Lita 4.6's chat service API for Slack-specific functionality. You can access this API object by calling the `Lita::Robot#chat_service`. See the API docs for `Lita::Adapters::Slack::ChatService` for details about the provided methods.

## Threading

lita-slack supports Slack threading by default, it will reply to messages in the same context as they were created by default.
This means it will reply in channel if a message was sent from the channel, or in thread if a message was sent from a thread.

You can prepend the first line of your message with the ellipsis character `…` to start a new thread.

## API documentation

The API documentation, useful for plugin authors, can be found for the latest gem release on [RubyDoc.info](http://www.rubydoc.info/gems/lita-slack)

## License

[MIT](http://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 6.0.1 (2023-03-14)
* Increment waiting times when rate-limited.
* Add option to avoid SSL verification in the slack websocket connection.

## 6.0.0 (2023-01-19)
* Use linear backoff when rate-limited. 

## 5.0.0 (2023-01-17)
* Non-paginated API calls obey Slack rate-limit rules, waiting when the HTTP status code is 429.
* users.list API call is now paginated.
* Paginated API calls limit has been increased 500 by default (100-200 is the recommended value and 1000 is the maximum value).

## 4.0.0 (2022-12-30)
* Upgrade dependencies: lita >= 4.8.
* Ignore .idea folder in .gitignore.

## 3.0.0 (2022-11-22)

* Sentry integration for API error responses.

## 2.1.1 (2022-08-28)

* Replaces deprecated rtm.start with rtm.connect

## 2.1.0 (2022-06-23)

Flywire fork

* fixes im.open API deprecation 




spec/spec_helper.rb
require 'simplecov'
require 'simplecov-cobertura'
require 'simplecov-html'

if ENV['COVERAGE']
  SimpleCov.formatters = SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::CoberturaFormatter,
    SimpleCov::Formatter::HTMLFormatter
  ])

  SimpleCov.start do
    enable_coverage :branch
    primary_coverage :branch
  end
end

require "shopify-lita-slack"
require "lita/rspec"

Lita.version_3_compatibility_mode = false

RSpec.configure do |config|
  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end
end

Lita.configure do |config|
  config.redis[:host] = ENV["REDIS_HOST"]
  config.redis[:port] = ENV["REDIS_PORT"]
end

spec/lita/adapters/slack/rtm_connection_spec.rb
require "spec_helper"

describe Lita::Adapters::Slack::RTMConnection, lita: true do
  def with_websocket(subject, queue)
    thread = Thread.new { subject.run(queue, ping: nil) }
    thread.abort_on_exception = true
    yield queue.pop
    subject.shut_down
    thread.join
  end

  subject { described_class.new(robot, config, rtm_start_response) }

  let(:api) { instance_double("Lita::Adapters::Slack::API") }
  let(:registry) { Lita::Registry.new }
  let(:robot) { Lita::Robot.new(registry) }
  let(:raw_user_data) { Hash.new }
  let(:channel) { Lita::Adapters::Slack::SlackChannel.new('C2147483705', 'general', 1360782804, 'U023BECGF', metadata) }
  let(:metadata) { Hash.new }

  let(:rtm_start_response) do
    Lita::Adapters::Slack::TeamData.new(
      [],
      Lita::Adapters::Slack::SlackUser.new('U12345678', 'carl', nil, raw_user_data),
      [Lita::Adapters::Slack::SlackUser.new('U12345678', 'carl', '', raw_user_data)],
      [channel],
      "wss://example.com/"
    )
  end
  let(:token) { 'abcd-1234567890-hWYd21AmMH2UHAkx29vb5c1Y' }
  let(:queue) { Queue.new }
  let(:proxy_url) { "http://example.com:3128" }
  let(:config) { Lita::Adapters::Slack.configuration_builder.build }

  before do
    config.token = token
  end

  describe ".build" do
    before do
      allow(Lita::Adapters::Slack::API).to receive(:new).with(config).and_return(api)
      allow(api).to receive(:rtm_start).and_return(rtm_start_response)
    end

    it "constructs a new RTMConnection with the results of rtm.start data" do
      expect(described_class.build(robot, config)).to be_an_instance_of(described_class)
    end

    it "creates users with the results of rtm.start data" do
      expect_any_instance_of(Lita::Adapters::Slack::RTMConnection).to receive(:fork).and_yield do
        expect(Lita.logger).to receive(:debug).with("Inserting 1 users")
        expect(Lita::Adapters::Slack::UserCreator).to receive(:create_users)
      end

      described_class.build(robot, config)
    end

    it "creates rooms with the results of rtm.start data" do
      expect_any_instance_of(Lita::Adapters::Slack::RTMConnection).to receive(:fork).and_yield do
        expect(Lita.logger).to receive(:debug).with("Inserting 1 channels")
        expect(Lita::Adapters::Slack::RoomCreator).to receive(:create_rooms)
      end

      described_class.build(robot, config)
    end

  end

  describe "#im_for" do
    before do
      allow(Lita::Adapters::Slack::API).to receive(:new).with(config).and_return(api)
      allow(
        Lita::Adapters::Slack::IMMapping
      ).to receive(:new).with(api, []).and_return(im_mapping)
      allow(im_mapping).to receive(:im_for).with('U12345678').and_return('D024BFF1M')
    end

    let(:im_mapping) { instance_double('Lita::Adapters::Slack::IMMapping') }

    it "delegates to the IMMapping" do
      with_websocket(subject, queue) do |websocket|
        expect(subject.im_for('U12345678')).to eq('D024BFF1M')
      end
    end
  end

  describe "#run" do
    let(:event) { double('Event', data: '{}') }
    let(:message_handler) { instance_double('Lita::Adapters::Slack::MessageHandler') }

    it "creates the WebSocket" do
      with_websocket(subject, queue) do |websocket|
        expect(websocket).to be_an_instance_of(Faye::WebSocket::Client)
      end
    end

    context "with a proxy server specified" do
      before do
        config.proxy = proxy_url
      end

      it "creates the WebSocket" do
        with_websocket(subject, queue) do |websocket|
          expect(websocket).to be_an_instance_of(Faye::WebSocket::Client)
        end
      end
    end

    it "dispatches incoming data to MessageHandler" do
      allow(Lita::Adapters::Slack::EventLoop).to receive(:defer).and_yield
      allow(Lita::Adapters::Slack::MessageHandler).to receive(:new).with(
        robot,
        'U12345678',
        {},
      ).and_return(message_handler)

      expect(message_handler).to receive(:handle)

      # Testing private methods directly is bad, but it's difficult to get
      # the timing right when testing it by emitting the "message" event on
      # the WebSocket.
      subject.send(:receive_message, event)
    end

    context "when the WebSocket is closed from outside" do
      it "shuts down the reactor" do
        with_websocket(subject, queue) do |websocket|
          sleep 2.0 # intermittent test
          websocket.close
          expect(EM.stopping?).to be_truthy
        end
      end
    end

  end

  describe "#send_messages" do
    let(:message_json) { MultiJson.dump(id: 1, type: 'message', text: 'hi', channel: channel_id) }
    let(:channel_id) { 'C024BE91L' }
    let(:websocket) { instance_double("Faye::WebSocket::Client") }

    before do
      # TODO: Don't stub what you don't own!
      allow(Faye::WebSocket::Client).to receive(:new).and_return(websocket)
      allow(websocket).to receive(:on)
      allow(websocket).to receive(:close)
      allow(Lita::Adapters::Slack::EventLoop).to receive(:defer).and_yield
    end

    it "writes messages to the WebSocket" do
      with_websocket(subject, queue) do |websocket|
        expect(websocket).to receive(:send).with(message_json)

        subject.send_messages(channel_id, ['hi'])
      end
    end

    it "raises an ArgumentError if the payload is too large" do
      with_websocket(subject, queue) do |websocket|
        expect do
          subject.send_messages(channel_id, ['x' * 16_001])
        end.to raise_error(ArgumentError)
      end
    end
  end
end


lib/lita/adapters/slack.rb
require 'lita/adapters/slack/chat_service'
require 'lita/adapters/slack/rtm_connection'
require 'net/http'

module Lita
  module Adapters
    # A Slack adapter for Lita.
    # @api private
    class Slack < Adapter
      # Required configuration attributes.
      config :token, type: String, required: true
      config :proxy, type: String
      config :parse, type: [String]
      config :link_names, type: [true, false]
      config :unfurl_links, type: [true, false]
      config :unfurl_media, type: [true, false]

      # Provides an object for Slack-specific features.
      def chat_service
        ChatService.new(config)
      end

      def mention_format(name)
        "@#{name}"
      end

      # Starts the connection.
      def run(&block)
        return if rtm_connection
        Lita.logger.debug("Starting to build RTM connection")
        @rtm_connection = RTMConnection.build(robot, config)

        Lita.logger.debug("Done building RTM connection")

        lita_ssl_verification = ENV.fetch("LITA_SLACK_WS_CONNECTION_VERIFY_SSL", "true") == "true"
        unless lita_ssl_verification
          Lita.logger.info("SLL connection is going to be verified by Net::HTTP and OpenSSL")
          verify_ssl_connection
        end

        rtm_connection.run(nil, { tls: { verify_peer: lita_ssl_verification }  }, &block)
      end



      # Returns UID(s) in an Array or String for:
      # Channels, MPIMs, IMs
      def roster(target)
        api = API.new(config)
        room_roster target.id, api
      end

      # @param [Array] messages list of String messages
      # messages starting with the ellipsis character will start a new thread
      def send_messages(target, messages)
        api = API.new(config)
        channel = channel_for(target)

        timestamp = target.timestamp if target.respond_to?(:timestamp)
        thread_ts = target.thread_ts if target.respond_to?(:thread_ts)

        if messages[0] && messages[0][0] == '…'
          thread_ts = timestamp
          messages[0] = messages[0][1..-1]
        end

        if thread_ts
          api.reply_in_thread(channel, messages, thread_ts)
        else
          api.send_messages(channel, messages)
        end
      end

      def set_topic(target, topic)
        channel = target.room
        Lita.logger.debug("Setting topic for channel #{channel}: #{topic}")
        API.new(config).set_topic(channel, topic)
      end

      def shut_down
        return unless rtm_connection

        rtm_connection.shut_down
        robot.trigger(:disconnected)
      end

      private

      attr_reader :rtm_connection

      def channel_for(target)
        if target.private_message?
          rtm_connection.im_for(target.user.id)
        else
          target.room
        end
      end

      def channel_roster(room_id, api)
        response = api.channels_info room_id
        response['channel']['members']
      end

      # Returns the members of a group, but only can do so if it's a member
      def group_roster(room_id, api)
        response = api.groups_list
        group = response['groups'].select { |hash| hash['id'].eql? room_id }.first
        group.nil? ? [] : group['members']
      end

      # Returns the members of a mpim, but only can do so if it's a member
      def mpim_roster(room_id, api)
        response = api.mpim_list
        mpim = response['groups'].select { |hash| hash['id'].eql? room_id }.first
        mpim.nil? ? [] : mpim['members']
      end

      # Returns the user of an im
      def im_roster(room_id, api)
        response = api.mpim_list
        im = response['ims'].select { |hash| hash['id'].eql? room_id }.first
        im.nil? ? '' : im['user']
      end

      def room_roster(room_id, api)
        case room_id
        when /^C/
          channel_roster room_id, api
        when /^G/
          # Groups & MPIMs have the same room ID pattern, check both if needed
          roster = group_roster room_id, api
          roster.empty? ? mpim_roster(room_id, api) : roster
        when /^D/
          im_roster room_id, api
        end
      end

      def verify_ssl_connection(host = "wss-primary.slack.com", max_retries = 10, wait_time = 5)
        ok = false
        ssl_retry = 0
        while !ok && ssl_retry < max_retries
          begin
            Net::HTTP.start(host, '443', use_ssl: true) { |http| http.peer_cert }
            Lita.logger.info("SSL connection is verified")
            ok = true
          rescue OpenSSL::SSL::SSLError => e
            Lita.logger.info("SSL connection is not verified. Retry #{ssl_retry+1}. Error: #{e.message}. Backtrace: #{e.backtrace}")
            raise if ssl_retry == max_retries
            ok = false
            sleep(wait_time)
          end
        end
      end
    end

    # Register Slack adapter to Lita
    Lita.register_adapter(:slack, Slack)
  end
end


lib/lita/adapters/slack/api.rb
require 'thread'
require 'faraday'

require 'lita/adapters/slack/team_data'
require 'lita/adapters/slack/slack_im'
require 'lita/adapters/slack/slack_user'
require 'lita/adapters/slack/slack_source'
require 'lita/adapters/slack/slack_channel'

module Lita
  module Adapters
    class Slack < Adapter
      # @api private
      class API
        DEFAULT_PAGE_SIZE = 1000
        DEFAULT_RATE_LIMITED_WAIT_TIME = 65
        DEFAULT_MAX_RATE_LIMITED_WAIT_TIME = 600

        def initialize(config, stubs = nil)
          @config = config
          @stubs = stubs
          @post_message_config = {}
          @post_message_config[:parse] = config.parse unless config.parse.nil?
          @post_message_config[:link_names] = config.link_names ? 1 : 0 unless config.link_names.nil?
          @post_message_config[:unfurl_links] = config.unfurl_links unless config.unfurl_links.nil?
          @post_message_config[:unfurl_media] = config.unfurl_media unless config.unfurl_media.nil?
        end

        def im_open(user_id)
          response_data = call_api('conversations.open', users: user_id)

          SlackIM.new(response_data["channel"]["id"], user_id)
        end

        def channels_info(channel_id)
          call_api("channels.info", channel: channel_id)
        end

        def channels_list(params: {})
          conversations_list(types: ["public_channel"], params: params)
        end

        def groups_list(params: {})
          response = conversations_list(types: ["private_channel"], params: params)
          response['groups'] = response['channels']
          response
        end

        def mpim_list(params: {})
          response = conversations_list(types: ["mpim"], params: params)
          response['groups'] = response['channels']
          response
        end

        def im_list(params: {})
          response = conversations_list(types: ["im"], params: params)
          response['ims'] = response['channels']
          response
        end

        def conversations_list(types: ["public_channel"], params: {})
          params.merge!({
            types: types.join(',')
          })
          params[:limit] = DEFAULT_PAGE_SIZE unless params[:limit]
          call_paginated_api(method: 'conversations.list', params: params, result_field: 'channels')
        end

        def call_paginated_api(method:, params:, result_field:)
          Lita.logger.debug("Start #{method} paginated API call with parameters #{params}. Callers: #{caller.join(', ')}")

          page = 1
          result = call_api(
            method,
            params,
            page
          )

          next_cursor = fetch_cursor(result)
          old_cursor = nil
          page += 1
          while !next_cursor.nil? && !next_cursor.empty? && next_cursor != old_cursor
            old_cursor = next_cursor
            params[:cursor] = next_cursor
            next_page = call_api(
              method,
              params,
              page
            )
            next_cursor = fetch_cursor(next_page)
            result[result_field] += next_page[result_field]
            Lita.logger.debug("#{method} API call page #{page} obtained successfully (#{next_page[result_field].size} items returned)")
            page += 1
          end
          result
        end

        def send_attachments(room_or_user, attachments)
          call_api(
            "chat.postMessage",
            as_user: true,
            channel: room_or_user.id,
            attachments: MultiJson.dump(attachments.map(&:to_hash)),
          )
        end

        def open_dialog(dialog, trigger_id)
          call_api(
            "dialog.open",
            dialog: MultiJson.dump(dialog),
            trigger_id: trigger_id,
          )
        end

        def send_messages(channel_id, messages)
          call_api(
            "chat.postMessage",
            **post_message_config,
            as_user: true,
            channel: channel_id,
            text: messages.join("\n"),
          )
        end

        def reply_in_thread(channel_id, messages, thread_ts)
          call_api(
            "chat.postMessage",
            as_user: true,
            channel: channel_id,
            text: messages.join("\n"),
            thread_ts: thread_ts
          )
        end

        def delete(channel, ts)
          call_api("chat.delete", channel: channel, ts: ts)
        end

        def update_attachments(channel, ts, attachments)
          call_api(
            "chat.update",
            channel: channel,
            ts: ts,
            attachments: MultiJson.dump(attachments.map(&:to_hash))
          )
        end

        def set_topic(channel, topic)
          call_api("channels.setTopic", channel: channel, topic: topic)
        end

        def rtm_start
          Lita.logger.debug("Starting `rtm_connect` method")
          connect_response_data = call_api("rtm.connect")
          Lita.logger.debug("Started building TeamData")

          users_list_response_data = nil
          conversations_response_data = nil

          threads = []
          threads << Thread.new {
            users_list_response_data = users_list
          }
          threads << Thread.new {
            conversations_response_data = conversations_list(params: {exclude_archived: true})
          }
          threads.each(&:join)

          channels = conversations_response_data['channels'].select do |conversation|
            conversation['is_channel']
          end
          groups = conversations_response_data['channels'].select do |conversation|
            conversation['is_group']
          end
          ims = conversations_response_data['channels'].select do |conversation|
            conversation['is_im']
          end

          Lita.logger.debug("Obtained #{users_list_response_data["members"].size} users")
          Lita.logger.debug("Obtained #{channels.size} channels")
          Lita.logger.debug("Obtained #{groups.size} groups")
          Lita.logger.debug("Obtained #{ims.size} ims")

          team_data = TeamData.new(
            SlackIM.from_data_array(ims),
            SlackUser.from_data(connect_response_data["self"]),
            SlackUser.from_data_array(users_list_response_data["members"]),
            SlackChannel.from_data_array(channels) + SlackChannel.from_data_array(groups),
            connect_response_data["url"],
          )
          Lita.logger.debug("Finished building TeamData")
          Lita.logger.debug("Finishing method `rtm_connect`")
          team_data
        end

        private

        attr_reader :stubs
        attr_reader :config
        attr_reader :post_message_config

        def wait_for_rate_limit(response, method, retry_count, page = nil)
          default_rate_limited_wait_time = ENV.fetch("DEFAULT_RATE_LIMITED_WAIT_TIME", DEFAULT_RATE_LIMITED_WAIT_TIME).to_i
          default_max_rate_limited_wait_time = ENV.fetch("DEFAULT_MAX_RATE_LIMITED_WAIT_TIME", DEFAULT_MAX_RATE_LIMITED_WAIT_TIME).to_i

          request_type = 'non-paginated'
          request_type = "page #{page}" if page
          base_sleep_amount = response.headers.fetch('retry-after', default_rate_limited_wait_time).to_i
          sleep_amount = base_sleep_amount * retry_count
          topped_sleep_amount = [default_max_rate_limited_wait_time, sleep_amount].min
          rate_limiting_warning_message = "Rate-limited in #{request_type} #{method} request, retry #{retry_count}, will wait #{topped_sleep_amount} seconds"
          Lita.logger.info(rate_limiting_warning_message)

          sleep(topped_sleep_amount)
          sleep_amount
        end

        def call_api(method, post_data = {}, page = nil)
          request_type = 'non-paginated'
          request_type = "page #{page}" if page

          Lita.logger.debug("Starting #{request_type} #{method} request. Callers: #{caller.join(', ')}")

          url = "https://slack.com/api/#{method}"

          response = connection.post(
            url,
            { token: config.token }.merge(post_data)
          )
          retry_count = 1

          Lita.logger.debug("Finished #{request_type} request retry #{retry_count} to Slack API #{method} with HTTP status #{response.status}")

          while response.status == 429
            waited_time = wait_for_rate_limit(response, method, retry_count, page)
            Lita.logger.info("#{method} #{request_type} request retry #{retry_count} rate-limited, waited #{waited_time} seconds")
            Lita.logger.debug("Waited request Response body #{response.body}")

            response = connection.post(
              url,
              { token: config.token }.merge(post_data)
            )
            retry_count += 1
            Lita.logger.debug("Finished #{method} #{request_type} request retry #{retry_count} with HTTP status #{response.status}")
          end

          Lita.logger.debug("Obtained from #{method} #{request_type} request retry #{retry_count} a HTTP status #{response.status}")

          data = parse_response(response, method)
          Lita.logger.debug("Finished #{method} #{request_type} request retry #{retry_count} response")

          if data["error"]
            error_message = "Slack API #{request_type} #{method} request retry #{retry_count} returned an error: #{data["error"]}."
            log_to_sentry(error_message, url, post_data)
            Lita.logger.error(error_message)
            raise error_message
          else
            Lita.logger.debug("Successful #{request_type} #{method} request retry #{retry_count}")
          end

          Lita.logger.debug("#{method} #{request_type} request retry #{retry_count}. HTTP Status: #{response.status}. Body: '#{response.body}'. Headers: #{response.headers}.")

          data['__RESPONSE__'] = response
          data
        end

        def connection
          if stubs
            Faraday.new { |faraday| faraday.adapter(:test, stubs) }
          else
            options = {}
            unless config.proxy.nil?
              options = { proxy: config.proxy }
            end
            Faraday.new(options)
          end
        end

        def parse_response(response, method)
          unless response.status == 429 || response.success?
            raise "Slack API call to #{method} failed with status code #{response.status}: '#{response.body}'. Headers: #{response.headers}"
          end

          MultiJson.load(response.body)
        end

        def fetch_cursor(page)
          page.dig("response_metadata", "next_cursor")
        end

        def users_list
          call_paginated_api(method: 'users.list', params: {limit: DEFAULT_PAGE_SIZE}, result_field: 'members')
        end

        def log_to_sentry(message, url, post_data)
          require 'sentry-ruby'
          Sentry.capture_exception(
            StandardError.new(message),
            tags: {
              build_sha: ENV.fetch("GITLAB_COMMIT_SHA", ""),
              version: File.read(".semver"),
              trace_id: "no_trace_id",
              user: "unknown",
              lita_cmd: url
            },
            extra: {
              lita_cmd: url,
              maintainers: "Tools, #dev-tools for support",
              post_data: post_data
            }
          )
        end
      end
    end
  end
end
